autoconnect/autoconnect-ws/autoconnect-ws-sm/src/identified/on_client_msg.rs

```diff
@@ -0,0 +1,204 @@
+
+    async fn _register(
+        &mut self,
+        channel_id: &Uuid,
+        key: Option<String>,
+    ) -> Result<String, SMError> {
+        if let Some(user) = &self.deferred_user_registration {
+            trace!(
+                "💬WebPushClient::register: User not yet registered... {:?}",
+                &user.uaid
+            );
+            self.app_state.db.add_user(user).await?;
+            self.deferred_user_registration = None;
+        }
+
+        let endpoint = make_endpoint(
+            &self.uaid,
+            channel_id,
+            key.as_deref(),
+            &self.app_state.endpoint_url,
+            &self.app_state.fernet,
+        )
+        .map_err(|e| SMError::MakeEndpoint(e.to_string()))?;
+        self.app_state
+            .db
+            .add_channel(&self.uaid, channel_id)
+            .await?;
+        Ok(endpoint)
+    }
+
+    async fn unregister(
+        &mut self,
+        channel_id: Uuid,
+        code: Option<u32>,
+    ) -> Result<ServerMessage, SMError> {
+        trace!("WebPushClient:unregister";
+               "uaid" => &self.uaid.to_string(),
+               "channel_id" => &channel_id.to_string(),
+               "code" => &code,
+        );
+        // TODO: (copied from previous state machine) unregister should check
+        // the format of channel_id like register does
+
+        let result = self
+            .app_state
+            .db
+            .remove_channel(&self.uaid, &channel_id)
+            .await;
+        let smsg = match result {
+            Ok(_) => {
+                self.app_state
+                    .metrics
+                    .incr_with_tags("ua.command.unregister")
+                    .with_tag("code", &code.unwrap_or(200).to_string())
+                    .send();
+                self.stats.unregisters += 1;
+                ServerMessage::Unregister {
+                    channel_id,
+                    status: 200,
+                }
+            }
+            Err(e) => {
+                error!("WebPushClient::unregister failed: {}", e);
+                ServerMessage::Unregister {
+                    channel_id,
+                    status: 500,
+                }
+            }
+        };
+        Ok(smsg)
+    }
+
+    async fn broadcast_subscribe(
+        &mut self,
+        _broadcasts: HashMap<String, String>,
+    ) -> Result<Option<ServerMessage>, SMError> {
+        unimplemented!();
+    }
+
+    async fn ack(&mut self, _updates: &[ClientAck]) -> Result<Vec<ServerMessage>, SMError> {
+        // TODO:
+        self.maybe_post_process_acks().await
+    }
+
+    async fn nack(&mut self, _code: Option<i32>) -> Result<(), SMError> {
+        unimplemented!();
+    }
+
+    async fn ping(&mut self) -> Result<ServerMessage, SMError> {
+        // TODO: why is this 45 vs the comment describing a minute?
+        // Clients shouldn't ping > than once per minute or we disconnect them
+        if sec_since_epoch() - self.last_ping >= 45 {
```

Comment:
If anything, this ought to be a config setting or at least a default. Feels mystery meat.

Reply:
👍 added to the TODO